### PR TITLE
EUREKA-240: add module descriptor validator

### PR DESCRIPTION
**Pull request by yauhen-vavilkin**
## Purpose

https://folio-org.atlassian.net/browse/EUREKA-240

To fix all Module Descriptors with an acceptance mechanism, there is a suggestion to create a Maven plugin. This plugin should be able to validate module descriptors during the build phase of the module and run as a separate process as well

## Approach

- add validators
- add tests

### CODE REVIEW

File: src/main/java/org/folio/md/validator/service/impl/PermissionDefinitionValidator.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package org.folio.md.validator.service.impl;
+
+import org.folio.md.validator.model.ValidationContext;
+import org.folio.md.validator.service.Validator;
+
+public class PermissionDefinitionValidator implements Validator {
+
+  @Override
+  public void validate(ValidationContext ctx) {
+    ctx.getModuleDescriptor().getProvides()
```
Comment (pfilippov-epam) [inline src/main/java/org/folio/md/validator/service/impl/PermissionDefinitionValidator.java:10]: possible NPE

File: src/main/java/org/folio/md/validator/service/impl/UniquePermissionForEndpointValidator.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import org.folio.md.validator.model.RoutingEntry;
+import org.folio.md.validator.model.ValidationContext;
+import org.folio.md.validator.service.Validator;
+
+public class UniquePermissionForEndpointValidator implements Validator {
+
+  @Override
+  public void validate(ValidationContext ctx) {
+    var moduleDescriptor = ctx.getModuleDescriptor();
+    moduleDescriptor.getProvides().stream()
```
Comment (pfilippov-epam) [inline src/main/java/org/folio/md/validator/service/impl/UniquePermissionForEndpointValidator.java:20]: possible NPE

File: src/main/java/org/folio/md/validator/ModuleDescriptorValidator.java [hunk 3]
```diff
@@ -1,30 +1,100 @@
...
+      getLog().warn(message);
+    }
+  }
+
+  private void handleFailure(String message, Throwable cause) throws MojoExecutionException {
+    if (failOnInvalidDescriptor) {
+      throw new MojoExecutionException(message, cause);
+    } else {
+      getLog().warn(message, cause);
+    }
```
Comment (pfilippov-epam) [inline src/main/java/org/folio/md/validator/ModuleDescriptorValidator.java:82]:
```suggestion
    if (failOnInvalidDescriptor) {
      throw new MojoExecutionException(message, cause);
    }
    getLog().warn(message, cause);
```

File: src/main/java/org/folio/md/validator/ModuleDescriptorValidator.java [hunk 4]
```diff
@@ -1,30 +1,100 @@
...
+      return null;
+    }
+  }
+
+  private void handleFailure(String message) throws MojoExecutionException {
+    if (failOnInvalidDescriptor) {
+      throw new MojoExecutionException(message);
+    } else {
+      getLog().warn(message);
+    }
```
Comment (pfilippov-epam) [inline src/main/java/org/folio/md/validator/ModuleDescriptorValidator.java:74]:
```suggestion
    if (failOnInvalidDescriptor) {
      throw new MojoExecutionException(message);
    }
    getLog().warn(message);
```

File: src/main/java/org/folio/md/validator/ModuleDescriptorValidator.java [hunk 5]
```diff
@@ -1,30 +1,100 @@
...
+  boolean failOnInvalidDescriptor;
+
+  private final ValidatorManager validator = initializeValidator();
+  private final ObjectMapper objectMapper = initializeObjectMapper();
 
   @Override
-  public void execute() throws MojoExecutionException, MojoFailureException {
+  public void execute() throws MojoExecutionException {
+    if (moduleDescriptroFile == null || !moduleDescriptroFile.exists()) {
+      handleFailure("Module descriptor file is not found");
```
Comment (pfilippov-epam) [inline src/main/java/org/folio/md/validator/ModuleDescriptorValidator.java:41]: check case, if file is not found with failOnInvalidDescriptor=false

File: src/main/java/org/folio/md/validator/service/impl/PermissionDefinitionValidator.java (hunk 1 above)
Comment (yauhen-vavilkin) [reply src/main/java/org/folio/md/validator/service/impl/PermissionDefinitionValidator.java]: moduleDescriptor cannot be null as well [as provides](https://github.com/folio-org/folio-module-descriptor-validator/blob/4b0fae0140b1f33b4ed90089dbc965a58bf3f0af/src/main/java/org/folio/md/validator/model/ModuleDescriptor.java#L22)

File: src/main/java/org/folio/md/validator/service/impl/UniquePermissionForEndpointValidator.java (hunk 2 above)
Comment (yauhen-vavilkin) [reply src/main/java/org/folio/md/validator/service/impl/UniquePermissionForEndpointValidator.java:20]: moduleDescriptor cannot be null as well [as provides](https://github.com/folio-org/folio-module-descriptor-validator/blob/4b0fae0140b1f33b4ed90089dbc965a58bf3f0af/src/main/java/org/folio/md/validator/model/ModuleDescriptor.java#L22)

**Review by pfilippov-epam — APPROVED** [review]

**Review by dmtkachenko — APPROVED** [review]

**Pull request merged**